WIP feature(relationships): adds function to query a set of relationships

mrclay opened the pull request:
Earlier you needed to first get entities from a relationship and then process relationships one by one for each entity. Now it is possible to query the relationships directly.

Fixes #7299

**feature(services): Adds a service for building and executing queries**

`elgg()->queries` has methods `select`, `update`, `update`, and `delete` for creating query builder objects usable for programmatically building SQL.

Instead of a result statement, the builder object's `execute` method returns the output of the corresponding Elgg database function, e.g. `get_data`, `update_data`, etc.

The builder extends Doctrine's to auto-prefix tables and create parameter sets for use with in `IN` operator.

Softly deprecates the "sanitize" functions in favor of the query builder.
- [ ] Add docs to elgg_get_relationships and for "queries" service
- [ ] break testCanFetchRelationships() into separate tests
- [ ] test for OrderBy
- [ ] tests for query builder acceptance in DB
- [ ] move DB additions into 1 commit

Commit:
feature(relationships): adds function to query a set of relationships

Earlier you needed to first get entities from a relationship and then process
relationships one by one for each entity. Now it is possible to query the
relationships directly.

Fixes #7299

feature(db): adds an SQL query builder based on Doctrine DBAL

Adds `elgg_get_query_builder()` and allows passing the builder object into
the core database functions like `get_data()`, etc.

The builder extends Doctrine's to auto-prefix tables and create parameter
sets for use with in `IN` operator.

Softly deprecates the "sanitize" functions in favor of the query builder.

mrclay commented:
QB provides a nice encapsulation of query & params, so we should probably expose this and nudge devs toward using it instead of raw SQL. It ties us to DBAL, but I think the win would be worth it.

juho-jaakkola commented on engine/classes/Elgg/Database/RelationshipsTable.php:
```diff
@@ -258,13 +311,15 @@ public function removeAll($guid, $relationship = "", $inverse_relationship = fal
-		$query = "SELECT * from {$this->db->getTablePrefix()}entity_relationships WHERE {$where}";
+		$qb = $this->db->getQueryBuilder()
+			->select('*')
+			->from($this->db->getTablePrefix() . 'entity_relationships')
```
Did you check whether query builder has built-in support for table prefixes?

mrclay replied:
It does not, but I just added it :)

mrclay commented on engine/classes/Elgg/Database.php:
```diff
@@ -126,6 +127,19 @@ protected function getConnection($type) {
+	 */
+	public function getQueryBuilder() {
+		// note we supply the readwrite connection, but when we execute in executeQueryBuilder()
+		// we will use the actual connection needed.
```
This could backfire if Elgg ever allows different connections to have different DB engines (with different SQL syntaxes). Ways around this I see:

- (preferred) Instead of getting a raw builder, have functions `elgg_build_select()`, `elgg_build_insert()`, et.al., which return the builder set up with the appropriate connection.
- Make user give the connection type (e.g. "read") to get the builder.

mrclay commented:
@Elgg/contributors this PR aims to expose Doctrine's query builder to help us move away from raw SQL. I'm thinking usage will end up like:
```php
$qb = elgg_build_select('r.*')
  ->from('{entity_relationships}', 'r')
  ->setMaxResults(10);
$rows = get_data($qb);
```

Thoughts? I'm not crazy about exposing the raw Connection, because it would be hard to wrap in Elgg's query cache, counters, logging, etc. Yes, Doctrine has a query cache, but for a long time most of our queries won't go through it and I doubt we don't want to maintain two caches.

mrclay committed:
feature(services): Adds a service for building and executing queries

`elgg()->queries` has methods `select`, `update`, `update`, and `delete` for
creating query builder objects usable for programmatically building SQL.

Instead of a result statement, the builder object's `execute` method returns
the output of the corresponding Elgg database function, e.g. `get_data`,
`update_data`, etc.

mrclay commented:
Latest API provides a dedicated service for getting builders (yay no added functions), the builders are preset for the 4 types, and the execute() method calls the corresponding Elgg query method.

```php
$qb = elgg()->queries->select('r.*')
  ->from('{entity_relationships}', 'r')
  ->setMaxResults(10);
$rows = $qb->execute($my_callback);
```

hypeJunction commented:
Do you have examples of more complex queries? Nested selects, multiple order clauses etc

mrclay commented:
https://www.google.com/search?q=dbal+query+builder
It's made to target the backends that Doctrine needs so it's a pretty generic solution. I've already added a method to append SQL for `ON DUPLICATE ..` statements. It can't and shouldn't replace some of our queries.

What I like is that it just encapsulates the SQL and parameters into one object which allows passing those down into our caching/timing infrastructure. We could trivially add a setSQL() method to set raw SQL for the whole query and still get the benefits of parameterization.

I guess the bigger Q is whether we should support direct use of regular DBAL (the standard PDO API). It has a cache built-in but we'd have to customize to work in callbacks and timing. Maybe we just don't support those in the short term.

hypeJunction commented:
I don't think I have an opinion on this. I will concur with whatever is decided.

mrclay commented:
I suggest we adopt this less radical API improvement in the short term: #9462.
